File: apps/mobile/ios/VideoSpliter/FFmpeg/VideoProcessorError.swift
//
//  VideoProcessorError.swift
//  VideoSpliter
//
//  Video processing error types
//

import Foundation

/**
 * Errors that can occur during video processing operations
 */
enum VideoProcessorError: Error, LocalizedError {
  case invalidVideoPath
  case invalidOutputDirectory
  case ffmpegExecutionFailed(String)
  case videoNotFound
  case unsupportedFormat
  case insufficientStorage
  case cancelled
  case unknown(Error)

  /**
   * Human-readable error description
   */
  var errorDescription: String? {
    switch self {
    case .invalidVideoPath:
      return "Invalid video path provided"
    case .invalidOutputDirectory:
      return "Invalid output directory"
    case .ffmpegExecutionFailed(let message):
      return "FFmpeg execution failed: \(message)"
    case .videoNotFound:
      return "Video file not found"
    case .unsupportedFormat:
      return "Unsupported video format"
    case .insufficientStorage:
      return "Insufficient storage space"
    case .cancelled:
      return "Operation was cancelled"
    case .unknown(let error):
      return "Unknown error: \(error.localizedDescription)"
    }
  }

  /**
   * Error code for JavaScript bridge
   */
  var code: String {
    switch self {
    case .invalidVideoPath: return "INVALID_VIDEO_PATH"
    case .invalidOutputDirectory: return "INVALID_OUTPUT_DIR"
    case .ffmpegExecutionFailed: return "FFMPEG_FAILED"
    case .videoNotFound: return "VIDEO_NOT_FOUND"
    case .unsupportedFormat: return "UNSUPPORTED_FORMAT"
    case .insufficientStorage: return "INSUFFICIENT_STORAGE"
    case .cancelled: return "CANCELLED"
    case .unknown: return "UNKNOWN"
    }
  }
}

File: apps/mobile/android/app/src/main/java/com/videospliter/videoprocessing/VideoProcessorError.kt
package com.videospliter.videoprocessing

/**
 * Error codes for video processing operations
 * Must match iOS VideoProcessorError and TypeScript VideoProcessingErrorCode exactly
 */
enum class VideoProcessorError(val code: String, val message: String) {
    INVALID_VIDEO_PATH("INVALID_VIDEO_PATH", "Video file path is invalid"),
    INVALID_OUTPUT_DIR("INVALID_OUTPUT_DIR", "Output directory is invalid or cannot be created"),
    FFMPEG_FAILED("FFMPEG_FAILED", "Video processing failed"),
    VIDEO_NOT_FOUND("VIDEO_NOT_FOUND", "Video file not found at the specified path"),
    UNSUPPORTED_FORMAT("UNSUPPORTED_FORMAT", "Unsupported or corrupted video format"),
    INSUFFICIENT_STORAGE("INSUFFICIENT_STORAGE", "Not enough storage space available"),
    CANCELLED("CANCELLED", "Operation was cancelled by user"),
    UNKNOWN("UNKNOWN", "An unknown error occurred");

    /**
     * Create an exception with this error
     */
    fun toException(details: String? = null): VideoProcessingException {
        val fullMessage = if (details != null) "$message: $details" else message
        return VideoProcessingException(code, fullMessage)
    }
}

/**
 * Custom exception for video processing errors
 */
class VideoProcessingException(
    val code: String,
    message: String,
    cause: Throwable? = null
) : Exception(message, cause)
